presto-main/src/main/java/com/facebook/presto/type/IpAddressOperators.java

```diff
@@ -0,0 +1,145 @@
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.lang.String.format;
+import static java.lang.System.arraycopy;
+
+public final class IpAddressOperators
+{
+    private IpAddressOperators()
+    {
+    }
+
+    @ScalarOperator(EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean equal(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.equals(right);
+    }
+
+    @ScalarOperator(NOT_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean notEqual(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return !left.equals(right);
+    }
+
+    @ScalarOperator(LESS_THAN)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean lessThan(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) < 0;
+    }
+
+    @ScalarOperator(LESS_THAN_OR_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean lessThanOrEqual(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) <= 0;
+    }
+
+    @ScalarOperator(GREATER_THAN)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean greaterThan(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) > 0;
+    }
+
+    @ScalarOperator(GREATER_THAN_OR_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean greaterThanOrEqual(@SqlType(StandardTypes.IPADDRESS) Slice left, @SqlType(StandardTypes.IPADDRESS) Slice right)
+    {
+        return left.compareTo(right) >= 0;
+    }
+
+    @ScalarOperator(BETWEEN)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean between(@SqlType(StandardTypes.IPADDRESS) Slice value, @SqlType(StandardTypes.IPADDRESS) Slice min, @SqlType(StandardTypes.IPADDRESS) Slice max)
+    {
+        return min.compareTo(value) <= 0 && value.compareTo(max) <= 0;
+    }
+
+    @ScalarOperator(HASH_CODE)
+    @SqlType(StandardTypes.BIGINT)
+    public static long hashCode(@SqlType(StandardTypes.IPADDRESS) Slice value)
+    {
+        return XxHash64.hash(value);
+    }
+
+    @LiteralParameters("x")
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.IPADDRESS)
+    public static Slice castFromVarchartoIpAddress(@SqlType("varchar(x)") Slice slice)
+    {
+        try {
+            byte[] address = forString(slice.toStringUtf8()).getAddress();
+            byte[] bytes = new byte[16];
+            if (address.length == 4) {
+                bytes[10] = (byte) 0xff;
+                bytes[11] = (byte) 0xff;
+                arraycopy(address, 0, bytes, 12, 4);
+            }
+            else if (address.length == IP_ADDRESS_SIZE) {
+                arraycopy(address, 0, bytes, 0, IP_ADDRESS_SIZE);
```

Comment:
No need to copy the resulting array, it is already a private copy. This can be
```java
byte[] bytes;
if (address.length == 4) {
    bytes = new byte[16];
    ...
}
else if (address.length == 16) {
    bytes = address;
}
```
Also, to copy an entire array, the idiomatic way is to use `clone()` (pretty much the only valid use of `clone()`)